### DISCUSSION TITLE
Numbers: Math mode design

### DISCUSSION THREAD

**Discussion by rcoreilly** [cogentcore/cogent#324] [Ideas]
This discussion is for spec of math mode in `numbers`, which builds on `cosh` [shell](https://github.com/cogentcore/core/tree/main/shell)

# Math mode

`cosh` establishes two modes: Go vs. `exec`, with explicit `{ }` = go, backticks = exec and implicit parsing rules to determine the mode.

Math mode is _always_ explicit, using `$ <expr> $`.

Hopefully, the inline form will be sufficient for integrating with standard  Go mode control flow, as in:
```go
for _, x := range $[1,2,3]$ {
    fmt.Println($x^2$)
}
```

If we need more powerful math-mode augmented control flow expressions, we could add an `$$ <stmt> $$` syntax.

As in `cosh`, it is **essential** that all math-mode expressions are just _transpiled_ into corresponding Go-syntax code, which can then either be compiled into Go programs, or run interactively using the yaegi interpreter.

# Two major domains: tensors and special math expressions

There are two major domains of special expression syntax: tensor (array) and special math expressions and functions, each of which is discussed in separate sections below.  In general the tensor is backed by the core [tensor](https://github.com/cogentcore/core/tree/main/tensor) package, and uses the `numpy` syntax where it is sensible.

Special math expressions include direct access to `cos`, `sin` etc functions, which are automatically transpiled to corresponding `math` package calls.  The other major issue is default numerical types (e.g., default to `float64`) but also support for `float32` where needed.

# Also possible: more extended functional programming and automatic GPU execution

The [Jax](https://github.com/google/jax) framework is the latest buzz in ML, which can speed up lots of code by adopting functional programming ideas that naturally apply to the GPU.  We could potentially leverage [gosl](https://github.com/cogentcore/core/tree/main/vgpu/gosl) to provide the same kind of functionality.  This would be a huge win.

**Comment by rcoreilly** [comment]
# Tensor

[Tensor](https://github.com/cogentcore/core/tree/main/tensor) provides a native row-major (RM) data structure that implements all of the underlying numpy-like capabilities. See [numpy for matlab users](https://numpy.org/devdocs/user/numpy-for-matlab-users.html) for a nice concise summary of numpy (and MATLAB) syntax.

## Literals

```go
// 1D:
[1,2,3]

// 2D:
[[1,2], [3,4]]
 =
[1,2]
[3,4]

// 3D:
[[[1, 2], [3, 4]], [[5, 6], [7, 8]]]
=
[1,2], [5,6]
[3,4], [7,8]
```

## Access

Two types of access: indexing with single indexes or ranges (sub-spaces), and boolean filtering (boolean mask).

As in Go, indexed sub-slices are references to the original (modify original if written), and you use explicit `copy` function to make a copy (which can clone, rather than direct 2 arg copy).

```go
a := [1,2,3,4,5]
b := [[1,2], [3,4]]
// [1,2]
// [3,4]

a[0] // 1
a[-1] // 5,  negative indexing from end: len(a)-1
a[2:4] // [3,4] = subspace

b[0, :] // [1,3] = first column
b[:, 1] // [3,4] = 2nd row
b[:,-1] // [3,4] = last row

a <= 3 // [true, true, true, false, false]  bit mask
a[a<3] // [1,2,3]  bit mask indexing
```

**Comment by rcoreilly** [comment]
# Scope and Naming

I'm increasingly thinking that it doesn't make sense to maintain two separate shell-like Go language transpiler systems, and we should just make one all-purpose language that does everything.  Essentially a Go version of [IPython](https://github.com/ipython), with numpy and shell commands builtin (except better of course, because it is Go syntax and we don't have to use the "magic" `%` prefix for everything).  Math and data analysis is very general and powerful, even if you're not specifically doing data analysis, and hopefully we can make it so it doesn't interfere with basic shell-level functionality.  See also #348 for shell / math synergy in the form of the `datafs` virtual filesystem.

Thus, a key step is to come up with a good name for this thing.  `cosh` doesn't really capture the full scope.

Some considerations:
* Because it builds on Go, it is tempting to work Go into the name, but that tends to lead to "wonky" concatenated names.
* In general, simple one-word, semi-random yet evocative, cool names are preferred over more more explicit descriptive names (e.g., "gonum" is an obvious option for "go" + "numbers" but, aside from being taken, it isn't very cool or evocative..)
* The kinds of things we want to evoke should be broadly computational, not overly-math-specific yet somehow math and data related.  e.g., `turing` would be an obvious choice, but is already taken https://en.wikipedia.org/wiki/Turing_(programming_language) -- it is probably very difficult to come up with something novel if using well-known math people and concepts etc.
* Obviously, any connection to cogent and core would be good.  Could we just create a lot of confusion + synergy by calling the language `core` ?  It really is core to everything.  Would have to rename the tool..  Alternatively, `cogent`.  It is a good name, and would not require renaming the core tool..  https://trustworthy.systems/projects/OLD/cogent/ exists but seems to be getting toward abandonware?
* https://en.wikipedia.org/wiki/Statistics has a bunch of names: Adlan is a cool one.  "Student" aka Gosset is a fun hero of stats (beer + stats) and Gosset starts with Go.

**Reply by rcoreilly** [reply]
soro is a possible Japanese term that sounds cool and has the right meaning it seems: https://jisho.org/search/%E7%AE%97%20%23kanji

**Reply by rcoreilly** [reply]
densoro is electronic calculator

**Reply by rcoreilly** [reply]
sounds a bit like "tensor" :)

**Reply by rcoreilly** [reply]
`slang` is the *perfect* name:
* s = scripting, shell, statistics, science, simple...
* "slang" is shorthand for the "formal" language of Go (transpiling)
* it is edgy, cool, and yet also self-descriptive -- rare combination

many others have appreciated the many possibilities of this name, but none seems to be dominant:
https://www.google.com/search?q=slang+programming+language

strangely enough, the most discussed appears to be a proprietary language used within goldman-sachs --weird!

**Reply by kkoreilly** [reply]
As we have learned time and time again over the development of Cogent Core, good programming names are descriptive, not clever. `cli` is better than `grease` and `tree` is better than `ki` etc. Therefore, I am opposed to the name `slang`.

**Reply by rcoreilly** [reply]
propose an alternative.  `cosh` is not an appropriate name for the scope of the language.  see all the points above -- this is a full scientific, statistics, data processing language, not a "bash"-like shell.  and `cosh` is not really a great name for lots of reasons.

**Reply by rcoreilly** [reply]
.sl is basically unclaimed.  it isn't listed in our base/fileinfo/mimetypes files and https://fileinfo.com/extension/sl reports some kind of software license. `slang` `.sl` is looking pretty good.. :)

**Comment by rcoreilly** [comment]
It seems unlikely that we're going to be able to transpile even simple expressions like $x+1$ properly without knowing what `x` is -- if it is a tensor vs a scalar, it behaves very differently.  Options:

* take a page from the Go playbook, and require that tensor objects be represented by _uppercase initial_ letters (`X+1`), whereas scalars are _lowercase initial_ `x+1`.  It is a strong convention in math that a matrix is represented with an uppercase letter (e.g., https://en.wikipedia.org/wiki/Matrix_(mathematics) ) and tensors are just n-dimensional matricies, so that seems reasonable.  Go does not typically use uppercase for _local_ variables, and almost all math expressions will be using local vars (initialized from the #348 datafs), but there is no actual harm in that -- they won't be exported.  The main issue here is that _vectors_ (1d tensors) are typically _not_ capitalized, but are typically bolded or written with an arrow on top.

* somehow try to infer and maintain type information about the variables, so we can do the right thing.  this might require using the type inference passes of the go compiler?

In any case, dealing with the unnamed *results* of expressions will require some amount of type inference, e.g., knowing that `X+1` is a tensor but `x+1` is a scalar, when this is then combined with other things in complex expressions...

It is also possible that if we're tracking the nature of the relevant variables, we could avoid the need for `$ $` and just auto-applying the relevant logic to any mathematical expression?  That would be a bonus.

**Comment by rcoreilly** [comment]
# Go level impl: Indexed Tensor

The workhorse for all math computations will be the `tensor.Indexed`.  IndexView of Table has proven its worth. The basic unit of computation is the Tensor, so we need it at that level.

We can share the Indexes between Table and Tensor, but the Indexed type itself which just has the indexes and the pointer to the underlying Tensor or Table cannot be shared.  

tensor.Sort and tensor.Filter operate on Indexed, and filter implements `A[a > 2]` kind of filtering expressions.  Indexes are more general than bitmasks and are the only supported view.

# Generic named functions

All computation boils down to applying a function to indexed tensor(s).  In a duck typed language like python, there is no problem doing this with any call signature, but that doesn't work in Go.  Can we devise a fully generic call signature?

Names are important for auto-naming the results of functions, e.g., in the Sim logging example, you want to be able to pass in a vararg list of stats functions to call, and then have it name the resulting data accordingly.  Standard functions get enums, but free-form string names are necessary for arbitrary user-defined functions. Do we have some kind of giant map of all such functions? and a giant enum of all standard cases?

The existing `stats` code (which is not very clean, to say the least) has the following "standard" signatures, typically with a fixed set of enum-defined implementations.

* `stats.StatFunc`: `func(idx int, val float64, agg float64) float64` -- it is unclear if the idx arg is ever used -- in principle it could tell you which item is the Min, etc, but we can't return multiple values..
    - e.g., `func SumFunc(idx int, val float64, agg float64) float64 { return agg + val }`

* `stats.Stat64(a []float64, stat Stats) float64` -- manual re-write of 32 and 64 bit for diff floats -- doesn't use the StatFunc :)
 
* `metric.Func64`:  `Func64 func(a, b []float64) float64`
    - e.g., `SumSquares64` has more complex scale, sumSquares agg values, followed by a computation, with some optimizations based on length, and checking for equal length etc.  could rewrite as a composite of agg iteration and final calc functions, but requires multiple return values from agg iterator.

* `norm.Func64`: `func(a []float64) float64`
* `DivNorm64(a []float64, nfunc Func64)`
    - e.g., `func ZScore64(a []float64) {  SubNorm64(a, stats.Mean64);  DivNorm64(a, stats.Std64) }`

Some of these operate on raw `[]float64` and `[]float32` values, and then other funcs operate on `Tensor`. 

The Tensor case needs to be able to operate on high-dimensional "Cell" values, e.g., the Mean stats on tensor returns a flat slice of float64 with the mean for each sub-tensor value..

Per above, we need to consolidate on an indexed tensor as the one data type to operate on universally.

## Element-wise vs. whole tensor functions

The simplest case is all functions operate element-wise, and you just call them with the `vmap` function that applies each elem func to the elements of the tensor(s).  Different functions can require diff number of tensors, and treat some as input and others as output.  Thus, a single vararg list of tensors is the most generic signature.

If you need to do some combination of element-wise and aggregate computation (as in the SumSquares in metric), you could aggregate into a 1-row tensor that holds the aggregate values, and then the aggregate function operates on those.

Can we establish a hard simplifying constraint that `tensor.Indexed` only indexes the outer-most "Row" (row major) index?  You can always get a subspace of a tensor to zoom in on a "Cell" component of row, and operate on that.  

However, in `axon`, our neuron-level computation functions operate on a "composite" index of Neuron * DataIndex (di), and the underlying memory image is `[Vars][Neurons][Data]` and want to be able to use a flexible function to map iteration index into element index.  

Thus, you need a function that maps the outer iteration index into an `[]int` of computed indexes, which are then interpreted by the element-wise function in a way that it expects..  Rather than codify that part specifically, we can allow the Element-wise function to interpret a bare idx input and then just need an "NFunc" to determine the total N to iterate over.

```Go
type ElemFunc func(idx int, tensors ...tensor.Indexed) // idx can be expanded by func into actual effective indexes

type NFunc func(tensors ...tensor.Indexed) int // gives the N to iterate over in `VMap`

func VMap(nfunc NFunc, elemFunc ElemFunc, tensors ...tensor.Indexed) {
    n := nfunc(tensors)
    for i := range n {
        elemFunc(i, tensors...)
    }
}
```

By making everything so generic, there is a significant possibility of runtime errors in calling the wrong combination of funcs and tensors, which could be mitigated by `gosl` doing some more powerful kind of validation..   This is the classic tradeoff of flexible weak typing vs. strong typing..  It is not clear how an error return might work in all of this.

Next step: write all of the existing stats functions in the above framework and see how it looks!

**Reply by rcoreilly** [reply]
By adopting the `tensor.Indexed` as the universal computational unit (even scalars get promoted to this), we retain the ability to transpile expressions in `cosl` without excessive type tracking -- everything is just converted into the corresponding sequence of function calls involving `tensor.Indexed` args.  Literals are separate tokens and can be wrapped, and every variable otherwise is the correct type.

`stats` is good, and has its own registered map of stats functions that operate on an input tensor and put results in an output tensor that has the outer row dim reduced to 1.

`tmath` will be a full wrapping of all `math` functions, mapping from in to out tensors of the same size.

**Reply by rcoreilly** [reply]
this is going to be key, but also disruptive:

* all tensor indexed access uses vararg instead of []int!  
* dim names are rare and set via a separate method.
* access is easy: `Float(i ...int)`
* set you just have to switch with the value first: `SetFloat(v, i...int)`
